libcaf_core/src/detail/rfc3629.cpp

```diff
@@ -49,57 +54,74 @@ bool validate_rfc3629(const std::byte* first, const std::byte* last) {
-      if (first == last || !is_continuation_byte(*first++))
-        return false;
+        return {checkpoint, false};
+      if (first == last)
+        return {checkpoint, true};
+      if (!is_continuation_byte(*first++))
+        return {checkpoint, false};
       continue;
     }
     // 1110'xxxx: 3-byte sequence.
     if (head<4>(x) == 0b1110'0000_b) {
-      if (first == last || !is_continuation_byte(*first))
-        return false;
+      if (first == last)
+        return {checkpoint, true};
+      if (!is_continuation_byte(*first))
+        return {checkpoint, false};
       // No non-shortest form.
       if (x == 0b1110'0000_b && head<3>(*first) == 0b1000'0000_b)
-        return false;
+        return {checkpoint, false};
       // No surrogate characters.
       if (x == 0b1110'1101_b && head<3>(*first) == 0b1010'0000_b)
-        return false;
+        return {checkpoint, false};
       ++first;
-      if (first == last || !is_continuation_byte(*first++))
-        return false;
+      if (first == last)
+        return {checkpoint, true};
+      if (!is_continuation_byte(*first++))
+        return {checkpoint, false};
       continue;
     }
     // 1111'0xxx: 4-byte sequence.
     if (head<5>(x) == 0b1111'0000_b) {
       // Check if code point is in the valid UTF range.
       if (x > std::byte{0xf4})
-        return false;
-      if (first == last || !is_continuation_byte(*first))
-        return false;
+        return {checkpoint, false};
+      if (first == last)
+        return {checkpoint, true};
+      if (!is_continuation_byte(*first))
+        return {checkpoint, false};
       // No non-shortest form.
       if (x == 0b1111'0000_b && head<4>(*first) == 0b1000'0000_b)
-        return false;
+        return {checkpoint, false};
       // Check if code point is in the valid UTF range.
       if (x == std::byte{0xf4} && *first >= std::byte{0x90})
-        return false;
+        return {checkpoint, false};
       ++first;
-      if (first == last || !is_continuation_byte(*first++))
-        return false;
-      if (first == last || !is_continuation_byte(*first++))
-        return false;
+      if (first == last)
+        return {checkpoint, true};
+      if (!is_continuation_byte(*first++))
+        return {checkpoint, false};
+      if (first == last)
+        return {checkpoint, true};
+      if (!is_continuation_byte(*first++))
+        return {checkpoint, false};
       continue;
     }
-    return false;
+    return {checkpoint, false};
   }
-  return true;
+  return {last, false};
 }
 
 } // namespace
 
 namespace caf::detail {
 
 bool rfc3629::valid(const_byte_span bytes) noexcept {
-  return validate_rfc3629(bytes.begin(), bytes.end());
+  return validate_rfc3629(bytes.begin(), bytes.end()).first == bytes.end();
+}
+
+std::pair<size_t, bool> rfc3629::validate(const_byte_span bytes) noexcept {
+  auto [last, incomplete] = validate_rfc3629(bytes.begin(), bytes.end());
+  return {last - bytes.begin(), incomplete};
```

Comment:
This silently converts a singed to an unsigned type; should be explicit:

```suggestion
  return {static_cast<size_t>(last - bytes.begin()), incomplete};
```